replay2/memory_remapper/typesafe_address.h

```diff
@@ -0,0 +1,36 @@
+// Copyright (C) 2022 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef REPLAY2_MEMORY_REMAPPER_TYPESAFE_ADDRESS_H
+#define REPLAY2_MEMORY_REMAPPER_TYPESAFE_ADDRESS_H
+
+namespace agi {
+namespace replay2 {
+
+class TypesafeAddress {
+   public:
+    explicit TypesafeAddress(char* address) : address_(address) {}
+    char* charPtr() const { return address_; }
```

Comment:
I guess it can be. It was most natural to use void* but the C++ spec doesn't allow you to do things like calculate offsets between void* pointers using the - operator. So I picked char as the basetype because it's a single byte type. Thinking harder maybe I should probably be using uint8_t here.

Reply:
My concern is when reading the memory "\0" may cause undesired behaviour due to char* and also it gives some sort of "string" expectation.

I don't know which version we use the C++ but afaik, byte added to C++ with 17. Otherwise maybe we can use uint8_t*

Reply:
Other than that for the const do you think should we return const for every pointer and reference unless we absolutely need to change them. For this one I believe we can have a set function to change if we need to.

Reply:
I've changed this to std::byte which I believe to the right thing. However, we may be remapping a capture address into a replay address so that we can write into that memory. For this reason I'm leaving the pointer as non-const.